ui/ducks/ens.js

```diff
@@ -0,0 +1,183 @@
+import { isHexString } from 'ethereumjs-util';
+
+import { getCurrentChainId } from '../selectors';
+import {
+  CHAIN_ID_TO_NETWORK_ID_MAP,
+  MAINNET_NETWORK_ID,
+} from '../../shared/constants/network';
+import {
+  CONFUSING_ENS_ERROR,
+  ENS_ILLEGAL_CHARACTER,
+  ENS_NOT_FOUND_ON_NETWORK,
+  ENS_NOT_SUPPORTED_ON_NETWORK,
+  ENS_NO_ADDRESS_FOR_NAME,
+  ENS_REGISTRATION_ERROR,
+  ENS_UNKNOWN_ERROR,
+} from '../pages/send/send.constants';
+import { isValidDomainName } from '../helpers/utils/util';
+import { CHAIN_CHANGED } from '../store/actionConstants';
+import {
+  BURN_ADDRESS,
+  isBurnAddress,
+  isValidHexAddress,
+} from '../../shared/modules/hexstring-utils';
+
+// Local Constants
+const ZERO_X_ERROR_ADDRESS = '0x';
+
+const initialState = {
+  stage: 'UNINITIALIZED',
+  resolution: null,
+  error: null,
+  warning: null,
+  network: null,
+};
+
+export const ensInitialState = initialState;
+
+const name = 'ENS';
+
+let ens = null;
+
+const slice = createSlice({
+  name,
+  initialState,
+  reducers: {
+    processEnsError: (state, action) => {
+      const { recipient, network, reason } = action.payload;
+      if (
+        isValidDomainName(recipient) &&
+        reason.message === 'ENS name not defined.'
+      ) {
+        state.error =
+          network === MAINNET_NETWORK_ID
+            ? ENS_NO_ADDRESS_FOR_NAME
+            : ENS_NOT_FOUND_ON_NETWORK;
+      } else if (reason.message === 'Illegal Character for ENS.') {
+        state.error = ENS_ILLEGAL_CHARACTER;
+      } else {
+        log.error(reason);
+        state.error = ENS_UNKNOWN_ERROR;
+      }
+    },
+    processEnsResult: (state, action) => {
+      state.resolution = action.payload;
+      if (state.resolution === BURN_ADDRESS) {
+        state.error = ENS_NO_ADDRESS_FOR_NAME;
+      }
+      if (state.resolution === ZERO_X_ERROR_ADDRESS) {
+        state.error = ENS_REGISTRATION_ERROR;
+      }
+      if (
+        isValidDomainName(state.resolution) &&
+        isConfusing(state.resolution)
+      ) {
+        state.warning = CONFUSING_ENS_ERROR;
+      }
+    },
+    enableEnsLookup: (state, action) => {
+      state.stage = 'INITIALIZED';
+      state.error = null;
+      state.resolution = null;
+      state.warning = null;
+      state.network = action.payload;
+    },
+    disableEnsLookup: (state) => {
+      state.stage = 'NO_NETWORK_SUPPORT';
+      state.error = ENS_NOT_SUPPORTED_ON_NETWORK;
+      state.warning = null;
+      state.resolution = null;
+      state.network = null;
+    },
+    resetResolution: (state) => {
+      state.resolution = null;
+      state.warning = null;
+      state.error =
+        state.stage === 'NO_NETWORK_SUPPORT'
+          ? ENS_NOT_SUPPORTED_ON_NETWORK
+          : null;
+    },
+  },
+  extraReducers: (builder) => {
+    builder.addCase(CHAIN_CHANGED, (state, action) => {
+      if (action.payload !== state.currentChainId) {
+        state.stage = 'UNINITIALIZED';
```

Comment:
We should be setting `ens` to `null` here as well I think? It certainly shouldn't be used after switching chains.

Reply:
yeah. Resolved in https://github.com/MetaMask/metamask-extension/pull/10965/commits/251ba67938f7758fd3d8ded3e27262d022c05a23